system-helper/flatpak-system-helper.c

```diff
@@ -1151,6 +1331,354 @@ handle_run_triggers (FlatpakSystemHelper   *object,
+          g_autoptr(GFile) repo_file = g_file_get_child (cache_dir, name);
+          g_autoptr(GFile) state = g_file_get_child (repo_file, "state");
+          g_autoptr(GFile) commit_partial = g_file_get_child (state, commit_partial_basename);
+
+           if (g_file_query_exists (commit_partial, NULL))
+            {
+              g_autoptr(OstreeRepo) repo = ostree_repo_new (repo_file);
+              g_autofree gchar *repo_path = g_file_get_path (repo_file);
+              /* Validate repo before returning */
+              if (!ostree_repo_open (repo, NULL, &error))
+                {
+                  g_warning ("Partial pull available at %s, but invalid repo: %s", repo_path, error->message);
+                  return FALSE;
+                }
+              else
+                {
+                  *out_repo_location = g_strdup (repo_path);
+                  return TRUE;
+                }
+            }
+        }
+    }
+
+   return FALSE;
+}
+
+static gboolean
+check_partial_pull_availability (const gchar *repo_tmp,
+                                 const gchar *ref_basename,
+                                 const gchar *commit,
+                                 gchar      **out_location)
+{
+  g_autoptr(GFileEnumerator) enumerator = NULL;
+  g_autoptr(GFile) repo_tmpfile = NULL;
+  GFileInfo *file_info = NULL;
+  g_autofree gchar *prefix = NULL;
+  g_autoptr(GError) error = NULL;
+  const gchar *name;
+
+  g_debug ("Checking partial pull availability for ref %s (commit: %s)", ref_basename, commit);
+
+  repo_tmpfile = g_file_new_for_path (repo_tmp);
+  enumerator = g_file_enumerate_children (repo_tmpfile,
+                                          G_FILE_ATTRIBUTE_STANDARD_NAME,
+                                          G_FILE_QUERY_INFO_NONE, NULL, NULL);
+  if (enumerator == NULL)
+    {
+      g_warning ("Failed to get file enumerator: %s", error->message);
+      return FALSE;
+    }
+
+   prefix = g_strdup_printf ("flatpak-cache-%s", ref_basename);
+
+   /* We are looking for flatpak-cache-$ref-XXXXXX directory in repo/tmp.
+      If we find one, we try to look for an OSTree repo which possibly has
+      the $commit.commitpartial file denoting the partial pull for that $commit. */
+  while (TRUE)
+    {
+      if (!g_file_enumerator_iterate (enumerator, &file_info, NULL, NULL, &error))
+        {
+          g_warning ("Error while file enumerator iterating: %s", error->message);
+          break;
+        }
+
+      if (file_info == NULL)
+        break;
+
+      name = g_file_info_get_name (file_info);
+      if (g_file_info_get_file_type (file_info) == G_FILE_TYPE_DIRECTORY &&
+          g_str_has_prefix (name, prefix) &&
+          !g_str_has_suffix (name, "mountpoint"))
+        {
+          g_autoptr(GFile) cache_dir = g_file_get_child (repo_tmpfile, name);
+          OngoingPull *pull;
+
+          /* If this cache directory is used by any other ongoing pull, we
+           * should not be re(using) it. */
```

Comment:
I think we need to be a bit more careful than this. Consider this chain of events:

1. client gets a revokefs FD to write into cache directory A
2. system helper crashes (due to some bug, the OOM killer killing it, etc)
3. client requests a revokefs FD for the same commit
4. system helper respawns (with `ongoing_pulls` empty) and returns a new revokefs FD to write to cache directory A
5. client calls Deploy for A
6. system helper kills the writer created in step 4, but the one from step 1 is still alive
7. now the client can write to root-owned files by racing with the chown + deploy

Is there some way we can ensure that any open writers get killed if the system helper goes away? A heavyweight way to do it would be to have the system helper create two pairs of sockets (P, Q) and (R, S), pass P to the writer and return S to the client, and splice Q and R together. Another way would be to keep a lock file on disk, deleted when the writer gets killed, and refuse to reuse the directory if the lock file is present. Perhaps there's a better solution than either of these.

Reply:
>Is there some way we can ensure that any open writers get killed if the system helper goes away?
There is an answer for this now. We add a new arg to `revokefs-fuse` i.e. `--exit-with-fd` which is the read end of a pipe. If the helper is killed, the `--backend` process responsible for writing on the disk will exit too. 
Keeping it unresolved to have further discussion (or you can go ahead and resolve it).

(apologies for late comment here. I forgot that this was still outstanding to write it down on the PR.)